mu/app.py

```diff
@@ -256,6 +258,36 @@ def setup_modes(editor, view):
     }
 
 
+class MutexError(BaseException):
+    pass
+
+
+class SharedMemory(object):
+
+    NAME = "mu-memory"
+
+    def __init__(self):
+        self._shared_memory = QSharedMemory(self.NAME)
+
+    def __enter__(self):
+        self._shared_memory.lock()
+        return self._shared_memory
+
+    def __exit__(self, *args, **kwargs):
+        self._shared_memory.unlock()
+
+    def acquire(self):
+        if self._shared_memory.attach():
+            pid = struct.unpack("l", self._shared_memory.data()[:4])
+            raise MutexError("MUTEX: Mu is already running with pid %d" % pid)
+        else:
+            self._shared_memory.create(8)
+            self._shared_memory.data()[:4] = struct.pack("l", os.getpid())
```

Comment:
Perhaps being over-cautious: Are we sure we are going to fit a pid in 4?

Reply:
Well it's a good point. I did briefly check the Win implementation of `os.getpid` but I didn't check the rest. We can easily up it, so I'll probably do that.

Reply:
I've just gone down a rabbit-hole I'd rather not have trying to work out from the Posix spec. what type is defined in terms of what. After I'd eventually reached Cloud Cuckoo Land the only thing I was sure about was that the Posix `pid_t` is a signed integer "whatever that means" [near-enough quoting the spec]. So I've gone for an 8-byte Long Long.

Reply:
If I'm reading my copy of 1003.1 2008 correctly,  `pid_t` cannot exceed the width of `long`, so on that basis `ctypes.sizeof(ctypes.c_long)` is probably the ‘correct’ answer - but 8 is probably a safe bet